File: spring-boot-template/src/main/java/tui/meta/challenge/quotes/service/QuoteServiceImpl.java
package tui.meta.challenge.quotes.service;

import lombok.RequiredArgsConstructor;
import org.springframework.dao.TransientDataAccessException;
import org.springframework.retry.annotation.Backoff;
import org.springframework.retry.annotation.Retryable;
import org.springframework.stereotype.Service;
import tui.meta.challenge.quotes.controller.exceptions.QuoteNotFoundException;
import tui.meta.challenge.quotes.controller.model.QuoteDto;
import tui.meta.challenge.quotes.mapping.QuoteEntityToDtoFunc;
import tui.meta.challenge.quotes.repository.QuoteRepository;

import java.util.List;

@Service
@RequiredArgsConstructor
public class QuoteServiceImpl implements QuoteService {
    private final QuoteRepository quoteRepository;
    private final QuoteEntityToDtoFunc quoteEntityToDtoFunc;

    @Override
    @Retryable(
            retryFor = { TransientDataAccessException.class },
            maxAttempts = 2,
            backoff = @Backoff(delay = 100)
    )
    public QuoteDto getSingleQuote(String id) {
        return quoteRepository.findById(id)
                .map(quoteEntityToDtoFunc)
                .orElseThrow(() -> new QuoteNotFoundException("quote with given id:" + id + " is missing"));
    }

    @Override
    @Retryable(
            retryFor = { TransientDataAccessException.class },
            maxAttempts = 2,
            backoff = @Backoff(delay = 100)
    )
    public List<QuoteDto> getQuotesByAuthor(String author) {
        String sanitizedAuthor = author.replaceAll("[^a-zA-Z0-9\\s]", "");
        return quoteRepository.getQuoteByAuthorFullTextSearch(sanitizedAuthor).stream().map(quoteEntityToDtoFunc).toList();
    }

    @Override
    @Retryable(
            retryFor = { TransientDataAccessException.class },
            maxAttempts = 2,
            backoff = @Backoff(delay = 100)
    )
    public List<QuoteDto> getAllQuotes() {
        return quoteRepository.findAll().stream().map(quoteEntityToDtoFunc).toList();
    }
}

File: micronaut-template/src/main/kotlin/tui/meta/challenge/service/QuoteServiceImpl.kt
package tui.meta.challenge.service

import jakarta.inject.Singleton
import tui.meta.challenge.controller.exceptions.QuoteNotFoundException
import tui.meta.challenge.controller.model.QuoteDto
import tui.meta.challenge.repository.QuoteRepository

@Singleton
class QuoteServiceImpl(
    private val quoteRepository: QuoteRepository
) : QuoteService {

    override fun getQuoteById(id: String): QuoteDto = quoteRepository.findById(id)
        .orElseThrow { QuoteNotFoundException("there is no such quote with id=${id}") }
        .let(::QuoteDto)

    override fun getAllQuotes(): List<QuoteDto> =
        quoteRepository.findAll().map(::QuoteDto)

    // can be improved because right now it's just exact match case
    override fun getQuotesByAuthor(author: String): List<QuoteDto> {
        val sanitizedAuthor = author.replace(REGEX, "")
        return quoteRepository.findByQuoteAuthor(sanitizedAuthor).map(::QuoteDto)
    }

    companion object {
        val REGEX = Regex("[^a-zA-Z0-9\\s]")
    }
}